### DISCUSSION TITLE
Window Logical Dates with "Rolling" SLAs - Bridging Event Systems and Schedule Intervals

### DISCUSSION THREAD

**Discussion by JonnyWaffles** [apache/airflow#21840] [General]
### Description

Hi team,

First I want to thank you for all that you do! This feature is better described by a business scenario

> I expect to execute a dag that process an event, N number times during a schedule window (quarterly, monthly, etc), I don't know or care when they are processed, but I want to be warned if I am nearing the window's interval end and the N number of executions have not yet been successful.

Naming this feature request is tricky, as is searching for it, so you may have already considered it. If so please point me in the right direction. Essentially, we set up a trigger system that executes unscheduled dags "manually" for each event type. Whenever the event occurs, an event specific dag is triggered to handle the event. But now, we need a way to enforce and notify if these "ad hoc" jobs are approaching the end of their generic window. For example, "I expect 3 executions for event A in a quarter". I don't think such a convention exists. An `ExternalTaskSensor` requires knowing the exact `logical_date` for your target dag or task, but we wouldn't know the exact time an event would occur.

Therefore, we created a custom `ExternalDagDateRangeSensor` that uses a date range instead of specific logical_dates, and polls between the current and next `data_interval_end` (ie during the run time of the parent monitoring dag) for a success count over threshold.

Then we set up a parent "monitoring" dag for each known interval (quarterly, monthly, weekly) and add `ExternalDagDateRangeSensor` tasks for each observed target dag with a timeout matching the interval's duration. I'd submit a PR, but our implementation described above is pretty wonky, but it kind of works for our needs.

- Is there a better way to enforce "loose" or "variable" SLAs like "may occur any time in a month but has to occur twice?"
- Is this something the community would be interested in?

Lastly, as I understand it you can't send out multiple SLA miss notifications as you get close and closer to the timeout window. For now we just allow users to set a timeout offset, ie enter SLA miss state 2 days before the window closes. A more robust solution may be to create additional sensors that resume where the previous failed and thus can "renter" the SLA miss state with a more severe alarms as the window closes!

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by potiuk** [comment]
Could you please describe wbat you want to propose in a paragraph or so ? A PR as well would be better. From the description I do not really understand what you want to propose. 

I will also convert it to discussion. We do not need issues PRs are more important and if you want to discuss stuff it's better to do it in "discussion" or - if it is more "significant" should be discussed in the devlist. 

From the description here I am not even able to figure out what is the scope of the change you propose - how big it is, what it is really about.

**Reply by JonnyWaffles** [reply]
Hi team, thanks @potiuk for making this a discussion thread.

I am trying to think of a more clear way to describe the intent. The business scenario is, "Process a file at least N times a month but we don't know when or how many times the file will land".

This is tricky to do in Airflow because a single successful execution completes the dag for the desired interval. The approach we're currently using is externally triggering the "child" processor dag whenever the file lands, and using a "monitor" dag with modified `ExternalTaskSensor` tasks to check for N number of child dag successes within the monitor's interval. 

The monitor dag is a bit tricky because it needs to monitor children in "real time", not after the monthly interval ends (by then it would be too late to warn that we did not see N number of file drops within the month).

Does this explanation help clarify the intent any?

**Reply by potiuk** [reply]
This use case is going to be handled by https://cwiki.apache.org/confluence/display/AIRFLOW/AIP-48+Data+Dependency+Management+and+Data+Driven+Scheduling

**Reply by JonnyWaffles** [reply]
Thanks! Reading through the proposal and I am concerned over the static data set identifier (ie `s3://my-connection@bucket/order_data`. In our scenario the identifier would be dynamic such that the partition key could be "any day in the target month". We would also need a way to specify a minimum number of desired executions, ie "run at least two executions within the month or panic!" This may be an awkward edge case and we can always build something specific for our use case within the new dataset framework, but others may also find this useful. Cheers.

**Reply by potiuk** [reply]
I think you could take a look after it is implemented and released in the current form and if you see you can improve it - you are always welcome to start a more detailed proposal in the devlist and PR(s) to implement them.
